### PR TITLE
chore: Migrate FAQ module to Room database

### LINKED ISSUE #1653: Migrate FAQ module to Room database
**Describe the chore**
Use Room database in the FAQ module.

Parent issue: #1348 

**Would you like to work on the issue?**
- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by ShridharGoel**
Fixes #1653 

Checklist:

- [ ] I have checked for PMD and check-style issues 
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream development branch.

Changes: 

Implemented Room database and shifted FAQ module to Room.

### CODE REVIEW

**Comment by ShridharGoel** [comment]
@iamareebjamal Please have a look. Items are not getting deleted from the database although everything seems proper.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java [hunk 1]
```diff
@@ -72,9 +68,8 @@ public Completable deleteFaq(long id) {
         }
 
         return faqApi.deleteFaq(id)
-            .doOnComplete(() -> repository
-                .delete(Faq.class, Faq_Table.id.eq(id))
-                .subscribe())
+            .doOnComplete(() -> OrgaRoomDatabase
+                .getInstance(context).faqDao().deleteFaq(id))
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: You are not subscribing

**Comment by iamareebjamal** [comment]
Target all room PRs to room branch please

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 1 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Changed to this but still not working:

```
faqApi.deleteFaq(id)
            .toObservable()
            .flatMapCompletable(var -> {
                OrgaRoomDatabase
                    .getInstance(context).faqDao().deleteFaq(id);
                return Completable.complete();
            })
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread());
```

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Obviously it won't

```
OrgaRoomDatabase
     .getInstance(context).faqDao().deleteFaq(id);
```

This is never executed because it is never subscribed. Return this. Why returning Completable.complete()

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Had tried that too but it didn't work. Although, on using doOnSubscribe() or doOnComplete() with `DELETE FROM Faq` query it works fine (Deletes all FAQs from database). But doesn't work with the current query. So I think maybe there's some problem with the `DELETE FROM Faq WHERE id = :id` query?

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: I think this is because the FAQ id is different from the id which is used as PrimaryKey

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Updated. Now it's working.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import io.reactivex.Observable;
+
+@Dao
+public interface FaqDao {
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    void insertFaq(Faq faq);
+
+    @Query("DELETE FROM Faq WHERE id = :id")
+    void deleteFaq(long id);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Change void to completable

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.List;
+
+import io.reactivex.Observable;
+
+@Dao
+public interface FaqDao {
+
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    void insertFaq(Faq faq);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Same

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java [hunk 4]
```diff
@@ -18,31 +19,27 @@
...
-            .withRateLimiterConfig("Faqs", rateLimiter)
-            .withDiskObservable(diskObservable)
-            .withNetworkObservable(networkObservable)
-            .build();
+        return faqApi.getFaqs(eventId)
+            .doOnNext(faqs -> OrgaRoomDatabase
+                .getInstance(context).faqDao().getAllFaqs(eventId))
+            .flatMapIterable(faqs -> faqs)
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread());
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: This changes the behaviour. It aleays loads FAQs from API

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java (hunk 2 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: But `DELETE` queries can return only `void` and `int`?

From the [documentation](https://developer.android.com/reference/android/arch/persistence/room/Query):

`UPDATE or DELETE queries can return void or int. If it is an int, the value is the number of rows affected by this query.`

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: No, it can handle Completable

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java (hunk 3 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Shouldn't we use `Observable<Faq>` here?

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java (hunk 2 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Okay :+1:

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Updated accordingly but DELETE query isn't working.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java (hunk 3 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqDao.java]: Updated and working.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 4 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Please suggest how it should be modified.

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Fetch from DB and use switchIfEmpty

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqApi.java [hunk 5]
```diff
@@ -16,7 +16,7 @@
     Observable<List<Faq>> getFaqs(@Path("id") long id);
 
     @POST("faqs")
-    Observable<Faq> postFaq(@Body Faq faq);
+    Completable postFaq(@Body Faq faq);
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqApi.java]: Why is this changed to Completable?

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java [hunk 6]
```diff
@@ -72,8 +66,8 @@ public Completable deleteFaq(long id) {
         }
 
         return faqApi.deleteFaq(id)
-            .doOnComplete(() -> repository
-                .delete(Faq.class, Faq_Table.id.eq(id))
+            .doOnComplete(() -> OrgaRoomDatabase
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Change to flatMapCompletable

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java [hunk 7]
```diff
@@ -1,11 +1,12 @@
 package com.eventyay.organizer.data.faq;
 
+import android.content.Context;
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: No import of android classes in Repository. Use dependency injection to inject Database

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 4 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]:
```
return OrgaRoomDatabase
            .getInstance(context).faqDao().getAllFaqs(eventId)
            .switchIfEmpty(faqApi.getFaqs(eventId))
            .flatMapIterable(faqs -> faqs)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread());
```

Does this look correct to you?

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Why do we need flatmap iterable?

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: And no. This won't insert the FAQs in the DB when loaded from API

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: > Why do we need flatmap iterable?

`switchIfEmpty()` will return a `Observable<List<Faq>>` while the function needs to return `Observable<Faq>`

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 7 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Updated.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 4 above)
Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Make this function return Observable<List<Faq>>

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: > And no. This won't insert the FAQs in the DB when loaded from API

But this isn't even displaying the FAQs from database, even after removing flatMapIterable() and change the function return type.
I've checked that the FAQs are being fetched properly from the database, just that the ProgressBar remains there and the list does not show.

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: OK will fetch tomorrow and see. Before that, please change to Observable<List<Faq>>. Because of DB layer observable, the stream will never end. So Observable<Faq> onComplete will never get called and it can never be shown in the UI

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: > And no. This won't insert the FAQs in the DB when loaded from API

And this would again require the use of flatMapIterable, won't it? To convert Observable<List<Faq>> to Observable<Faq> for passing into the insert function.

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: > OK will fetch tomorrow and see. Before that, please change to Observable<List>. Because of DB layer observable, the stream will never end. So Observable onComplete will never get called and it can never be shown in the UI

Done.

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: No, don't convert to Observable<Faq>

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Okay, then we'll have to loop through the list and insert items into the DB?

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: No. https://developer.android.com/reference/androidx/room/Insert.html

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Okay we can insert the list directly

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Updated.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqApi.java (hunk 5 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqApi.java]: Reverted.

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 6 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Done

File: app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java (hunk 4 above)
Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: @iamareebjamal Did you check this?

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Is it still not working after the change?

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: No

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Logs:

```
android.database.sqlite.SQLiteConstraintException: UNIQUE constraint failed: resources.resourceid (code 1555)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:782)
        at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:788)
        at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:86)
        at android.database.sqlite.SQLiteDatabase.insertWithOnConflict(SQLiteDatabase.java:1474)
        at android.database.sqlite.SQLiteDatabase.insert(SQLiteDatabase.java:1343)
```

Comment (iamareebjamal) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: Well, it can't insert the FAQs. Add OnConflictStrategy in Insert annotation

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/data/faq/FaqRepositoryImpl.java]: It's already there.

File: app/src/main/java/com/eventyay/organizer/core/faq/list/FaqListPresenter.java [hunk 8]
```diff
@@ -33,56 +28,43 @@
...
         selectedMap.clear();
-        faqChangeListener.stopListening();
     }
 
     public void loadFaqs(boolean forceReload) {
         getFaqSource(forceReload)
             .compose(dispose(getDisposable()))
             .compose(progressiveErroneousRefresh(getView(), forceReload))
+            .flatMapIterable(faqs -> faqs)
             .toList()
```
Comment (iamareebjamal) [inline app/src/main/java/com/eventyay/organizer/core/faq/list/FaqListPresenter.java]: Remove last 2 lines

Comment (ShridharGoel) [reply app/src/main/java/com/eventyay/organizer/core/faq/list/FaqListPresenter.java]: Will have to create a new function in ViewTransformers then, with `ObservableTransformer` instead of `SingleTransformer`. Is that fine?

**Comment by ShridharGoel** [comment]
After migrating to MVVM and making a few changes, this is working fine. Will push the changes after #1659 is merged.

**Comment by mariobehling** [comment]
> #1659

It is merged. Please resolve conflicts here.

**Comment by iamareebjamal** [comment]
It's showing 20 commits

**Comment by ShridharGoel** [comment]
@iamareebjamal Squashed the commits.
